bridges/snowbridge/primitives/router/src/inbound/mod.rs

```diff
@@ -289,6 +313,49 @@ where
 			[GlobalConsensus(network), AccountKey20 { network: None, key: token.into() }],
 		)
 	}
+
+	fn convert_transact(
+		chain_id: u64,
+		sender: H160,
+		origin_kind: OriginKind,
+		fee: u128,
+		weight_at_most: Weight,
+		payload: Vec<u8>,
+	) -> (Xcm<()>, Balance) {
+		let xcm_fee: Asset = (Location::parent(), fee).into();
+
+		let xcm: Xcm<()> = vec![
+			// Only our inbound-queue pallet is allowed to invoke `UniversalOrigin`
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			// Change origin to the bridge.
+			UniversalOrigin(GlobalConsensus(Ethereum { chain_id })),
+			// DescendOrigin to the sender.
+			DescendOrigin(AccountKey20 { network: None, key: sender.into() }.into()),
+			// Withdraw fees from sender to pay the xcm execution
+			WithdrawAsset(xcm_fee.clone().into()),
+			// Pay for execution.
+			BuyExecution { fees: xcm_fee.clone(), weight_limit: Unlimited },
+			SetAppendix(Xcm(vec![
+				RefundSurplus,
+				// Deposit surplus back to sender.
+				DepositAsset {
+					assets: Wild(AllCounted(1u32)),
+					beneficiary: Location {
+						parents: 0,
+						interior: Junctions::from([
+							GlobalConsensus(Ethereum { chain_id }),
+							AccountKey20 { network: None, key: sender.into() },
+						]),
+					},
+				},
+			])),
+			// Transact on dest chain.
+			Transact { origin_kind, require_weight_at_most: weight_at_most, call: payload.into() },
+		]
+		.into();
+
+		(xcm, fee.into())
```

Comment:
If `fee` returns 0 then this [line for burning fees](https://github.com/Snowfork/polkadot-sdk/pull/114/files#diff-832917c2fd6302a9f175210f7d9941ea82a1a7c5a35aa35e662380074d405203L285) can be made simple. The fee returned from this function was intended to be the fee portion required to be burnt. The caller has access to the fee in the decoded message if needed.

Reply:
Not sure fully understand the point here. Actually the fee for transact is set as none zero by the sender and will also be paid by the sender on dest chain.